### PR TITLE
[Conpty] Pass through request for mouse mode to the Terminal

### PR DESCRIPTION
**Pull request by PankajBhojwani**
## Summary of the Pull Request
When the client application sets the console input mode with the `ENABLE_MOUSE_INPUT` flag, we send along the appropriate VT sequences to the connected terminal (if there is one).

 
## References
#376 
#6859 

## PR Checklist
* [ ] Closes #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [x] I work here

## Validation Steps Performed
Far manager works

### CODE REVIEW

File: src/host/getset.cpp
Comment (PankajBhojwani) [inline src/host/getset.cpp:380]: Only calling these 2 seems to make far manager work, but I am not sure if we should be sending other VT sequences through as well/in different cases (like only button events versus all mouse events), since the console mode only has 1 umbrella `ENABLE_MOUSE_INPUT` mode and not more specific ones as far as I know

File: src/host/inputBuffer.hpp
Comment (PankajBhojwani) [inline src/host/inputBuffer.hpp:82]: We probably want more functions for other VT sequences we may want to send

**Comment by PankajBhojwani** [comment]
Here's something interesting - tested this with far manager in terminal and it works fine with cmd, but there's a bug when using powershell: upon exiting far in powershell, for some reason powershell never asks us to set the input mode back to 'no mouse input', so Terminal is stuck in enable_mouse_input mode, breaking things like selection. I am not sure if this is a powershell issue or something that we need to deal with?

File: src/host/getset.cpp
Comment (skyline75489) [inline src/host/getset.cpp:382]: Am I reading this correctly? The things in the `else` clause are exactly the same as above...

File: src/host/getset.cpp
Comment (skyline75489) [reply src/host/getset.cpp:380]: Can you try if Win32 Vim works? The lacking of mouse support in Win32 Vim is the root cause of several issues. It also drives me to NeoVim personally.

File: src/host/getset.cpp
Comment (zadjii-msft) [inline src/host/getset.cpp:377]: should we do something like the `if (gci.GetInsertMode() != PreviousInsertMode)` lines above, where we only emit these sequences when they _changed_ `ENABLE_MOUSE_INPUT`? Might be less chatty

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp:382]: _oops_, thanks for spotting that!

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp:377]: That makes sense!

File: src/host/getset.cpp
Comment (DHowett) [inline src/host/getset.cpp:370]: You can probably simplify this pretty well:

```
const auto oldHasMouseMode{ WI_IsFlagSet(context.InputMode, ...) };
const auto newHasMouseMode{ WI_IsFlagSet(mode, ...) };
if (oldHasMouseMode != newHasMouseMode)
{
    PassThroughButtonEvent(newHasMouseMode);
}
```

File: src/host/getset.cpp
Comment (DHowett) [inline src/host/getset.cpp:368]: I'd suggest that we just combine them all into one function. "PassThroughWin32MouseRequest" which enables or disables all requisite modes (since there's no reason to have all two or three in the input buffer interface)

File: src/host/getset.cpp
Comment (DHowett) [reply src/host/getset.cpp]: We probably want 1003 instead of 1002 -- that's the "any event" mode which includes hover (the win32 event model automatically reports hover!)

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp:370]: Thanks for the suggestion!

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp:368]: Yep makes sense

File: src/host/getset.cpp
Comment (DHowett) [inline src/host/getset.cpp:372]: I'm not sure this is quite right -- we only ever care about the current final status of quickedit when mouse mode is being toggled, right?

What we want to know is "*will user input generate mouse records in conhost*?"

You can probably simplify the logic here substantially:

```
const auto oldMouseMode{ !oldQuickEditMode && WI_IsFlagSet(old, mouse) };
const auto newMouseMode{ !newQuickEditMode && WI_IsFlagSet(new, mouse) };
```

"The mouse was enabled prior if quickedit was off and mouse was on" "the mouse is newly-enabled if quickedit is off and mouse is on".

File: src/host/getset.cpp
Comment (DHowett) [inline src/host/getset.cpp:376]: It is okay to send the enable request multiple times :smile: it doesn't do any harm to the terminal, and it still only needs to be turned OFF one time.

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp:372]: Ah yes this way is much less verbose, fixed!

File: src/host/getset.cpp
Comment (PankajBhojwani) [reply src/host/getset.cpp]: Yeah, I just wanted to send as few repeated sequences as possible!

File: src/host/getset.cpp
Comment (DHowett) [inline src/host/getset.cpp:362]: we chatted about this checking the CONSOLE_QUICK_EDIT variation :)

Comment (PankajBhojwani) [reply src/host/getset.cpp]: Yep! Changed

**Comment by PankajBhojwani** [comment]
The issues with powershell and WSL have been resolved (as far as I can tell), so I've marked this PR ready for review now. 

However, certain applications (like vim) will still not work because they do not turn quick edit mode off.

**Comment by DHowett** [comment]
@msftbot make sure @zadjii-msft and @miniksa sign off

**Comment by DHowett** [comment]
I want @zadjii-msft to sign off because he probably has an opinion about InputBuffer having hardcoded VT sequences in it, and I want @miniksa to sign off because he probably has an opinion about whether this is the correct thing to do in general. Both Mikes are equipped to have that second opinion, but since this is closer to the conhost side... you know. :smile:

File: src/host/inputBuffer.cpp
Comment (DHowett) [inline src/host/inputBuffer.cpp:235]: You could probably put these in one request! Since we support DECSET with multiple parameters...

```suggestion
            LOG_IF_FAILED(_pTtyConnection->WriteTerminalW(L"\x1b[?1003;1006h"));
```

(I checked xterm- it looks like it supports multiple DECSET/DECRST modes in one mode line? I can't corroborate that with the VT510 manual. @j4james likely knows if this is standards-compliant... he explicitly made it work in #7799!)

Comment (j4james) [reply src/host/inputBuffer.cpp:235]: Yeah, this is standards compliant. It's documented in the DEC STD-070 manual, section 2.4.5.1 Set Mode.
> The SM control provides a means of placing one or more of the terminal mode values into the set state... The parameter list sent with a single instance of this control should not exceed the maximum allowable for the parser implementation. 

They recommend support for up to 16 parameters, but some terminals will have a lower limit than that, and I think there are also some that don't support multiple mode parameters at all. Unfortunately I don't have access to my terminal collection at the moment, so I'm not sure how big a problem that is.

Comment (DHowett) [reply src/host/inputBuffer.cpp:235]: Excellent, thank you

**Comment by zadjii-msft** [comment]
> certain applications (like vim) will still not work because they do not turn off quick edit mode.

wait then how the heck does mouse _ever_ work in `vim.exe`? Does it require users to manually turn quick edit off?

**Comment by DHowett** [comment]
> Does it require users to manually turn quick edit off?

IT SURE DOES

Check out the dates on [this mailing list post](https://groups.google.com/g/vim_dev/c/bQ7jfMwa8Zg) (view in a private browsing tab so Google's panopticon can't force you to log in.)

**Comment by zadjii-msft** [comment]
Wow, what a history lesson. Well, I suppose that's on the `vim` maintainers then?


or is the hot take that conpty should have QE disabled by default (since QE _doesn't really make sense in conpty)?

File: src/host/inputBuffer.cpp
Comment (PankajBhojwani) [reply src/host/inputBuffer.cpp:235]: Thank you! We only write once now

File: src/host/getset.cpp
Comment (zadjii-msft) [inline src/host/getset.cpp:362]: I'd almost add a comment with a reference to "GH#9970" here, because I doubt that a future reader would know why the heck QE has to be disabled to enable mouse mode. This thread has become the canonical source of truth on the matter, so it'll be a helpful reference for any future readers.

**Review by zadjii-msft — APPROVED** [review]
Yea that's a nit. Just don't want the notes to get lost if we ever move the repo again and break the `git blame` in 20 years.

**Comment by zadjii-msft** [comment]
hot take 2: should _we_ make a PR to `vim.exe` to have it disable QE mode? 

cold take 3: We should add a troubleshooting paragraph to the docs page that explains
* why mouse mode might not work in some win32 exes, 
* how the maintainers of said software can fix that
* A list of known applications that won't work right. Like, link to the vim.exe issue tracker for the request to add support (if such an feature tracker exists)

**Review by miniksa — APPROVED** [review]
Yeah OK. It's still mildly confusing to understand why this works just from the code itself... but the link back to this pull I guess is as good as we can do here. We've definitely done stuff like this before, so I don't have any reason to block it if it helps compatibility broadly.

**Comment by nabeelomer** [comment]
Hi, do you have development builds where I can try this out?

**Comment by wez** [comment]
re: vim.exe quickedit mode; I understand that vim doesn't turn it off, but!
Why does it default to on in terminal preview (and any app that runs openconsole.exe via conpty.dll) when the registry is set to disable it?
Is there a recommended way for a terminal emulator to turn it off by default?

**Comment by DHowett** [comment]
All fair questions.

ConPTY ignores any registry settings, even for the console defaults, because it's intended to provide a console interface that acts the same on all machines regardless of user configuration (so that the hosting terminal can make the final call about a number of things).

It then follows that it defaults to on because that's what it does in an unconfigured conhost :smile:

The problem now, though, is this:

* `ENABLE_MOUSE_INPUT` is _on by default_ (because that's also how it's always been), even for things like PowerShell and CMD (!) which do not use mouse events.
* Quick edit suppresses mouse mode.
* The only way to get mouse events in a Win32 console application is to have quick edit off and mouse mode on.
* If we switch `ENABLE_QUICK_EDIT_MODE` _off_ by default, or allow it as an option, the first sequence you'll get will probably be `\e[?1003;1006h`. You'll be in mouse mode all the time.
   * Won't this make it difficult for a terminal hosting a console application to determine when it should handle mouse events locally (scroll the buffer, select text, open a context menu)? Legitimate question -- I'd be worried about terminal emulators opting into a default experience that makes them _more like_ conhost. The whole point of ConPTY was to let people be _less_ like conhost, and not have to deal with conhost's baggage. :smile:

**Comment by wez** [comment]
Makes sense! tricky!
In the meantime, I went ahead and submitted a PR to vim to teach it to disable quickedit mode

**Comment by Sebastian-byte** [comment]
> Makes sense! tricky!
> In the meantime, I went ahead and submitted a PR to vim to teach it to disable quickedit mode

And it got merged!, now Win32 Vim supports mouse input.

**Pull request merged**
